File: backend/loop.py
"""
Loop = the orchestrator that turns a one-shot LLM rewrite (relative, unreliable)
into a standardized, on-target result (absolute, reproducible).

    source --> [Rewriter] --> draft --> [Referee measures] --> in target?
                  ^                                                |
                  |____________ feedback (measured grade) <________| (if no)

Also runs an optional fidelity check so you can see whether simplification is
dropping information. Every run is logged to a CSV row so you can A/B test
configs (model / prompt / target / scorer) against how classes actually go.
"""

from __future__ import annotations
import csv, os, time
from dataclasses import dataclass, field, asdict
from typing import Optional

from referee import get_scorer, MultiFormulaScorer, VocabScorer
from target import GradeBandTarget


@dataclass
class Round:
    n: int
    text: str
    grade: float
    detail: dict
    passed: bool
    feedback: str


@dataclass
class LevelingResult:
    source_grade: float
    final_grade: float
    rounds: list
    passed: bool
    config: dict
    fidelity: Optional[dict] = None

    @property
    def n_rounds(self) -> int:
        return len(self.rounds)


def run_loop(source_text: str,
             rewriter,
             target: GradeBandTarget,
             scorer=None,
             vocab_scorer=None,
             max_rounds: int = 4,
             verbose: bool = True) -> LevelingResult:
    """Core measure-and-iterate loop. `rewriter` is any object with
    .rewrite(text, target_desc, feedback) -> str."""
    scorer = scorer or MultiFormulaScorer()
    vocab_scorer = vocab_scorer or VocabScorer()

    src = scorer.score(source_text)
    if verbose:
        print(f"SOURCE: {src}")
        print(f"TARGET: {target.description}\n")

    rounds: list[Round] = []
    feedback = None
    current = source_text

    for i in range(1, max_rounds + 1):
        draft = rewriter.rewrite(source_text, target.description, feedback)
        sc = scorer.score(draft)
        vc = vocab_scorer.score(draft)
        passed, fb = target.check(sc.grade, pct_difficult=vc.detail["pct_difficult_words"])
        rounds.append(Round(i, draft, sc.grade, sc.detail, passed, fb))
        if verbose:
            print(f"--- round {i} ---")
            print(f"  measured: {sc}")
            print(f"  vocab: {vc.detail['pct_difficult_words']:.0f}% difficult words")
            print(f"  verdict: {fb}\n")
        current = draft
        feedback = fb
        if passed:
            break

    return LevelingResult(
        source_grade=src.grade,
        final_grade=rounds[-1].grade,
        rounds=rounds,
        passed=rounds[-1].passed,
        config={"rewriter": getattr(rewriter, "name", "?"),
                "scorer": scorer.name,
                "target": f"{target.low}-{target.high}",
                "max_rounds": max_rounds},
    )


def fidelity_check(source: str, rewrite: str, judge_rewriter=None) -> dict:
    """Optional: how much meaning survived? Cheap proxy + room for an LLM judge.
    The proxy is crude (content-word overlap); for real validation, plug an LLM
    judge in via judge_rewriter or a dedicated comprehension-question step."""
    import re
    def content_words(t):
        stop = set("the a an and or but of to in on at for with as is are was were "
                   "be been being it its this that these those by from we you they "
                   "he she i not no can will would could should has have had do does "
                   "did so if then than which who whom whose".split())
        return {w for w in re.findall(r"[a-z]+", t.lower()) if w not in stop and len(w) > 2}
    s, r = content_words(source), content_words(rewrite)
    kept = len(s & r) / max(len(s), 1)
    return {"content_word_retention": round(kept, 3),
            "source_unique_terms": len(s),
            "rewrite_unique_terms": len(r),
            "note": "Crude lexical proxy. Low retention can be fine (synonyms) "
                    "or bad (dropped facts) -- use an LLM judge for real fidelity."}


def log_result(result: LevelingResult, path: str, label: str = ""):
    """Append one row per run so you can compare configs over a semester."""
    row = {
        "timestamp": time.strftime("%Y-%m-%d %H:%M:%S"),
        "label": label,
        "rewriter": result.config["rewriter"],
        "scorer": result.config["scorer"],
        "target": result.config["target"],
        "source_grade": round(result.source_grade, 2),
        "final_grade": round(result.final_grade, 2),
        "n_rounds": result.n_rounds,
        "passed": result.passed,
        "fidelity_retention": (result.fidelity or {}).get("content_word_retention", ""),
    }
    new = not os.path.exists(path)
    with open(path, "a", newline="") as f:
        w = csv.DictWriter(f, fieldnames=list(row))
        if new:
            w.writeheader()
        w.writerow(row)
    return path

File: frontend/src/lib/loop.ts
import { multiFormulaScore, vocabScore, type ScoreResult } from "./readability";
import { checkTarget, describeTarget, type GradeBandTarget } from "./target";
import { rewrite, type Provider } from "./rewriters";

export type Round = {
  n: number;
  text: string;
  grade: number;
  detail: Record<string, number>;
  passed: boolean;
  feedback: string;
};

export type LevelingResult = {
  sourceGrade: number;
  finalGrade: number;
  passed: boolean;
  rounds: Round[];
  finalText: string;
  fidelity: { contentWordRetention: number; sourceUniqueTerms: number; rewriteUniqueTerms: number };
  config: { provider: Provider; model: string; target: string };
};

export type RunLoopOpts = {
  text: string;
  target: GradeBandTarget;
  provider: Provider;
  apiKey: string;
  model?: string;
  maxRounds: number;
  scorer?: (text: string) => ScoreResult;
  onRound?: (r: Round) => void;
};

export async function runLoop(opts: RunLoopOpts): Promise<LevelingResult> {
  const scorer = opts.scorer || multiFormulaScore;
  const source = scorer(opts.text);
  const targetDesc = describeTarget(opts.target);

  const rounds: Round[] = [];
  let feedback: string | null = null;

  for (let i = 1; i <= opts.maxRounds; i++) {
    const draft = await rewrite(
      opts.provider,
      opts.text,
      targetDesc,
      feedback,
      opts.apiKey,
      opts.model,
    );
    const sc = scorer(draft);
    const vc = vocabScore(draft);
    const { passed, feedback: fb } = checkTarget(opts.target, sc.grade, vc.pctDifficultWords);
    const round: Round = {
      n: i,
      text: draft,
      grade: sc.grade,
      detail: sc.detail,
      passed,
      feedback: fb,
    };
    rounds.push(round);
    opts.onRound?.(round);
    feedback = fb;
    if (passed) break;
  }

  const last = rounds[rounds.length - 1];
  return {
    sourceGrade: source.grade,
    finalGrade: last.grade,
    passed: last.passed,
    rounds,
    finalText: last.text,
    fidelity: fidelityCheck(opts.text, last.text),
    config: {
      provider: opts.provider,
      model: opts.model || "default",
      target: `${opts.target.low}-${opts.target.high}`,
    },
  };
}

const STOP = new Set(
  ("the a an and or but of to in on at for with as is are was were be been being it its " +
    "this that these those by from we you they he she i not no can will would could should " +
    "has have had do does did so if then than which who whom whose"
  ).split(" "),
);

function contentWords(t: string): Set<string> {
  const out = new Set<string>();
  for (const m of t.toLowerCase().matchAll(/[a-z]+/g)) {
    const w = m[0];
    if (w.length > 2 && !STOP.has(w)) out.add(w);
  }
  return out;
}

export function fidelityCheck(source: string, rewriteText: string) {
  const s = contentWords(source);
  const r = contentWords(rewriteText);
  let overlap = 0;
  for (const w of s) if (r.has(w)) overlap++;
  return {
    contentWordRetention: Math.round((overlap / Math.max(s.size, 1)) * 1000) / 1000,
    sourceUniqueTerms: s.size,
    rewriteUniqueTerms: r.size,
  };
}
